File: Reproduce/2025-TSP-IceBerg/Fig6_a.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Fig. 6(a) reproduction for
"Uncovering the Iceberg in the Sea: Fundamentals of Pulse Shaping and
Modulation Design for Random ISAC Signals".

The programming logic follows Fig2.py. The RRC pulse is replaced by the
PSL-designed pulse obtained from (44)--(49), as in Fig. 6(d).
"""

import numpy as np
import cvxpy as cp
import matplotlib.pyplot as plt
from Modulations import modulator


plt.rcParams["font.family"] = "Times New Roman"
plt.rcParams["font.size"] = 18
plt.rcParams["axes.titlesize"] = 18
plt.rcParams["axes.labelsize"] = 18
plt.rcParams["xtick.labelsize"] = 18
plt.rcParams["ytick.labelsize"] = 18
plt.rcParams["axes.unicode_minus"] = False
plt.rcParams["figure.figsize"] = [8, 6]
plt.rcParams["lines.linestyle"] = "-"
plt.rcParams["lines.linewidth"] = 2
plt.rcParams["lines.color"] = "blue"
plt.rcParams["lines.markersize"] = 6
plt.rcParams["figure.facecolor"] = "white"
plt.rcParams["axes.edgecolor"] = "black"
plt.rcParams["legend.fontsize"] = 14

np.random.seed(42)


def FFTmatrix(L):
    mat = np.zeros((L, L), dtype=complex)
    for i in range(L):
        for j in range(L):
            mat[i, j] = np.exp(-1j*2*np.pi*i*j/L)/np.sqrt(L)
    return mat


def solve_iceberg_shaping_psl(N, L, alpha, K_s1):
    """PSL iceberg-shaping problem in (44) and (49)."""

    N_alpha = int(alpha*N)
    N_non_rolloff = N-N_alpha
    N_zeros = N_non_rolloff//2
    N_ones = N_non_rolloff//2

    g = cp.Variable(N, nonneg=True)

    constraints = []

    # Reversed FFT-bin ordering used by the numerical implementation:
    # 1 -> monotonically decreasing roll-off -> 0.
    constraints.append(g[0:N_zeros] == 1)
    constraints.append(g[N-N_ones:N] == 0)

    for n in range(N-1):
        constraints.append(g[n+1]-g[n] <= 0)

    constraints.append(cp.sum(g) == N/2)

    psl_terms = []

    for k in K_s1:
        f_k = np.exp(-1j*2*np.pi*k*np.arange(N)/(L*N))
        gk = g+(1-g)*np.exp(-1j*2*np.pi*k/L)
        psl_terms.append(cp.abs(f_k.conj().T @ gk)**2)

    objective = cp.Minimize(cp.max(cp.hstack(psl_terms)))
    problem = cp.Problem(objective, constraints)
    problem.solve(verbose=False)

    if problem.status in [cp.OPTIMAL, cp.OPTIMAL_INACCURATE]:
        print("Optimization successful!")
        print(f"Optimal PSL value: {problem.value}")
        return g.value

    raise RuntimeError(f"Optimization failed, status: {problem.status}")


# %% Parameters in Fig. 6(a)
Tsym = 1
pi = np.pi
N = 128
L = 10
alpha = 0.35
kappa = 1.32

FLN = FFTmatrix(L*N)
FN = FFTmatrix(N)


# %% Generate the optimized pulse according to Fig. 6(d)
# The paper interval [5,15] uses symbol-delay units. The discrete ACF is
# sampled at Ts=T/L, and therefore uses k=5L,...,15L.
K_s1 = np.arange(5*L, 15*L + 1)

gN = solve_iceberg_shaping_psl(N, L, alpha, K_s1)

print("Optimal spectrum coefficients:")
print(gN)

g_N = 1-gN
g_design = np.hstack((gN, np.zeros((L-2)*N), g_N))

# g_design=N*|F_{LN}p|^2. Choose zero spectral phase to recover one
# valid pulse with exactly the optimized squared spectrum.
P_spectrum = np.sqrt(np.maximum(np.real(g_design), 0)/N)
p = FLN.conj().T @ P_spectrum
p = p/np.sqrt(np.sum(np.abs(p)**2))

norm2p = np.linalg.norm(p)
g = N*(FLN @ p)*(FLN.conj() @ p.conj())
g = np.real(g)

pulseSpectrumError = np.linalg.norm(g-g_design)/max(
    np.linalg.norm(g_design), np.finfo(float).eps
)
print(
    "Relative squared-spectrum reconstruction error: "
    f"{pulseSpectrumError:.3e}"
)


# %% OFDM theoretical average squared ACF, Eq. (36)
U = FN.conj().T
V = np.eye(N)
tilde_V = V*V.conj()

TheoAveACF_Iceberg = np.zeros(L*N)
TheoAveACF_OFDM_M1 = np.zeros(L*N)
TheoAveACF_OFDM_M10000 = np.zeros(L*N)

for k in range(L*N):
    gk = g[:N]+(1-g[:N])*np.exp(-1j*2*pi*k/L)
    f_k = np.exp(-1j*2*pi*k*np.arange(N)/(L*N))

    r1 = np.abs(gk @ f_k.conj())**2
    TheoAveACF_Iceberg[k] = r1

    M = 1
    r2 = (kappa-1)/M*(N - 2*(1-np.cos(2*pi*k/L))*np.sum(g[:N]*(1-g[:N])))
    TheoAveACF_OFDM_M1[k] = r1+r2

    M = 10000
    r2 = (kappa-1)/M*(N - 2*(1-np.cos(2*pi*k/L))*np.sum(g[:N]*(1-g[:N])))
    TheoAveACF_OFDM_M10000[k] = r1+r2

TheoAveACF_Iceberg = (TheoAveACF_Iceberg/TheoAveACF_Iceberg.max()+1e-14)
TheoAveACF_Iceberg = np.fft.fftshift(TheoAveACF_Iceberg)

TheoAveACF_OFDM_M1 = (TheoAveACF_OFDM_M1/TheoAveACF_OFDM_M1.max()+1e-14)
TheoAveACF_OFDM_M1 = np.fft.fftshift(TheoAveACF_OFDM_M1)

TheoAveACF_OFDM_M10000 = (TheoAveACF_OFDM_M10000/TheoAveACF_OFDM_M10000.max()+1e-14)
TheoAveACF_OFDM_M10000 = np.fft.fftshift(TheoAveACF_OFDM_M10000)


# %% Numerical average squared ACF, inherited from Fig2.py Eq. (26)
MOD_TYPE = "qam"
Order = 16
modem, Es, bps = modulator(MOD_TYPE, Order)
Constellation = modem.constellation/np.sqrt(Es)
AvgEnergy = np.mean(np.abs(Constellation)**2)

# Keep the full three-dimensional coherent-integration array. Fig. 6(a)
# displays one numerical realization of the 10,000-slot coherent result.
Iter = 100


# %% No coherent integration: M=1
M = 1
SimAveACF_OFDM_M1 = np.zeros((Iter, L*N))

for k in range(L*N):
    fk = FLN[:, k]
    fk_tilde = fk[:N]
    gk = g[:N]+(1-g[:N])*np.exp(-1j*2*pi*k/L)

    for it in range(Iter):
        d = np.random.randint(Order, size=N)
        s = Constellation[d]
        VHs = np.abs(V.conj().T @ s)**2
        SimAveACF_OFDM_M1[it, k] = np.abs(np.sum(gk*VHs*fk_tilde.conj()))**2


Sim_M1_avg = SimAveACF_OFDM_M1.mean(axis=0)
Sim_M1_avg = Sim_M1_avg/Sim_M1_avg.max()+1e-14
Sim_M1_avg = np.fft.fftshift(Sim_M1_avg)


# %% 10,000 coherent integrations: M=10000
M = 10000
SimAveACF_OFDM_M10000 = np.zeros((M, Iter, L*N), dtype=complex)

for k in range(L*N):
    fk = FLN[:, k]
    fk_tilde = fk[:N]
    gk = g[:N]+(1-g[:N])*np.exp(-1j*2*pi*k/L)

    for m in range(M):
        for it in range(Iter):
            d = np.random.randint(Order, size=N)
            s = Constellation[d]
            VHs = np.abs(V.conj().T @ s)**2

            # Do not square here. Average the complex ACF over M first.
            SimAveACF_OFDM_M10000[m, it, k] = np.sum(gk*VHs*fk_tilde.conj())


# %% Coherent averaging, Eq. (33)
RkBar = SimAveACF_OFDM_M10000.mean(axis=0)
RkBar2 = np.abs(RkBar)**2

Sim_M10000_avg = RkBar2.mean(axis=0)
Sim_M10000_avg = Sim_M10000_avg/Sim_M10000_avg.max()+1e-14
Sim_M10000_avg = np.fft.fftshift(Sim_M10000_avg)


# %% Plot Fig. 6(a)
x = np.arange(-N*L//2, N*L//2)

fig, axs = plt.subplots(1, 1, figsize=(10, 8), constrained_layout=True)

axs.plot( x, 10*np.log10(Sim_M1_avg), color="tab:blue", linestyle="-", linewidth=1.2, label="No Integration, Numerical" )
axs.plot( x, 10*np.log10(TheoAveACF_OFDM_M1), color="tab:blue", linestyle=":", linewidth=1.8, label="No Integration, Theoretical" )

axs.plot(x, 10*np.log10(Sim_M10000_avg), color="tab:orange", linestyle="-", linewidth=1.2, label="10k Coh Integration, Numerical")
axs.plot(x, 10*np.log10(TheoAveACF_OFDM_M10000), color="tab:orange", linestyle=":", linewidth=1.8, label="10k Coh Integration, Theoretical" )

axs.plot(x, 10*np.log10(TheoAveACF_Iceberg), color="black", linestyle="--", linewidth=1.2, label='"Iceberg" of the Designed Pulse')

axs.set_xlabel("Delay Index")
axs.set_ylabel("Ambiguity Level (dB)")
axs.set_xlim([-300, 300])
axs.set_ylim([-120, 0])
axs.set_xticks(np.arange(-300, 301, 100))
axs.set_yticks(np.arange(-120, 1, 20))
axs.grid(True)
axs.legend(loc="lower center", edgecolor="black")

out_fig = plt.gcf()
# out_fig.savefig("Fig6_a.png", dpi=300)
out_fig.savefig("Fig6_a.pdf")
plt.show()
plt.close()

File: Reproduce/2025-TSP-IceBerg/Fig6_a.m
%% Fig. 6(a)
% Uncovering the Iceberg in the Sea: Fundamentals of Pulse Shaping and
% Modulation Design for Random ISAC Signals
%
% OFDM, 16-QAM, N = 128, L = 10, alpha = 0.35.
% The pulse is first designed through the PSL problem in (44)--(49), and
% then used in the theoretical and numerical ACF calculations inherited
% from Fig2.py.
%
% CVX is required: http://cvxr.com/cvx/

clear;
close all;
clc;
rng(42,'twister');

set(groot,'defaultAxesFontName','Times New Roman');
set(groot,'defaultTextFontName','Times New Roman');
set(groot,'defaultAxesFontSize',18);
set(groot,'defaultTextFontSize',18);
set(groot,'defaultLineLineWidth',2);
set(groot,'defaultLineMarkerSize',6);
set(groot,'defaultFigureColor','white');
set(groot,'defaultLegendFontSize',14);

%% Parameters in Fig. 6(a)
Tsym = 1;
pi_value = pi;
N = 128;
L = 10;
alpha = 0.35;
kappa = 1.32;

FLN = FFTmatrix(L*N);
FN = FFTmatrix(N);

%% Generate the optimized pulse according to Fig. 6(d)
% The delay region [5,15] is expressed in symbol-delay units, whereas the
% discrete ACF is sampled at Ts = T/L.
K_s1 = 5*L:15*L;
gN = solve_iceberg_shaping_psl(N,L,alpha,K_s1);

g_N = 1-gN;
g_design = [gN;zeros((L-2)*N,1);g_N];

% g_design = N*|F_{LN}p|^2. A zero spectral phase is selected to recover
% one valid pulse having exactly the optimized squared spectrum.
P_spectrum = sqrt(max(real(g_design),0)/N);
p = FLN'*P_spectrum;
p = p/sqrt(sum(abs(p).^2));
 
g = N*(FLN*p).*(conj(FLN)*conj(p));
g = real(g);

pulseSpectrumError = norm(g-g_design)/max(norm(g_design),eps);
fprintf('Relative squared-spectrum reconstruction error: %.3e\n', pulseSpectrumError);

%% OFDM theoretical average squared ACF, Eq. (36)
U = FN';  
V = eye(N);
tilde_V = V.*conj(V);  

TheoAveACF_Iceberg = zeros(L*N,1);
TheoAveACF_OFDM_M1 = zeros(L*N,1);
TheoAveACF_OFDM_M10000 = zeros(L*N,1);

for k = 0:L*N-1
    gk = g(1:N)+(1-g(1:N))*exp(-1j*2*pi_value*k/L);
    fk = exp(-1j*2*pi_value*k*(0:N-1).'/(L*N));

    r1 = abs(gk.'*conj(fk))^2;
    TheoAveACF_Iceberg(k+1) = r1;

    M = 1;
    r2 = (kappa-1)/M* (N-2*(1-cos(2*pi_value*k/L))*sum(g(1:N).*(1-g(1:N))));
    TheoAveACF_OFDM_M1(k+1) = r1+r2;

    M = 1000;
    r2 = (kappa-1)/M* (N-2*(1-cos(2*pi_value*k/L))*sum(g(1:N).*(1-g(1:N))));
    TheoAveACF_OFDM_M10000(k+1) = r1+r2;
end

TheoAveACF_Iceberg = TheoAveACF_Iceberg/max(TheoAveACF_Iceberg)+1e-14;
TheoAveACF_Iceberg = fftshift(TheoAveACF_Iceberg);

TheoAveACF_OFDM_M1 = TheoAveACF_OFDM_M1/max(TheoAveACF_OFDM_M1)+1e-14;
TheoAveACF_OFDM_M1 = fftshift(TheoAveACF_OFDM_M1);

TheoAveACF_OFDM_M10000 = TheoAveACF_OFDM_M10000/max(TheoAveACF_OFDM_M10000)+1e-14;
TheoAveACF_OFDM_M10000 = fftshift(TheoAveACF_OFDM_M10000);

%% Numerical average squared ACF, inherited from Fig2.py Eq. (26)
MOD_TYPE = 'qam';  
Order = 16;
Constellation = qammod((0:Order-1).',Order,'gray','UnitAveragePower',true);
AvgEnergy = mean(abs(Constellation).^2); 

% Fig. 6(a) displays numerical realizations. Keep the full 3-D coherent-
% integration array, with one outer realization of the 10,000-slot result.
Iter = 1000;

%% No coherent integration: M = 1
M = 1;
SimAveACF_OFDM_M1 = zeros(Iter,L*N);

for k = 0:L*N-1
    fk = FLN(:,k+1);
    fk_tilde = fk(1:N);
    gk = g(1:N)+(1-g(1:N))*exp(-1j*2*pi_value*k/L);

    for it = 1:Iter
        d = randi([0,Order-1],N,1);
        s = Constellation(d+1);
        VHs = abs(V'*s).^2;
        SimAveACF_OFDM_M1(it,k+1) = abs(sum(gk.*VHs.*conj(fk_tilde)))^2;
    end
end

Sim_M1_avg = mean(SimAveACF_OFDM_M1,1);
Sim_M1_avg = Sim_M1_avg/max(Sim_M1_avg)+1e-14;
Sim_M1_avg = fftshift(Sim_M1_avg);

%% 10,000 coherent integrations: M = 10000
M = 1000;
SimAveACF_OFDM_M10000 = complex(zeros(M,Iter,L*N));

for k = 0:L*N-1
    fk = FLN(:,k+1);
    fk_tilde = fk(1:N);
    gk = g(1:N)+(1-g(1:N))*exp(-1j*2*pi_value*k/L);

    for m = 1:M
        for it = 1:Iter
            d = randi([0,Order-1],N,1);
            s = Constellation(d+1);
            VHs = abs(V'*s).^2;

            % Do not square here. Average the complex ACF over M first.
            SimAveACF_OFDM_M10000(m,it,k+1) = sum(gk.*VHs.*conj(fk_tilde));
        end
    end
end

%% Coherent averaging, Eq. (33)
RkBar = mean(SimAveACF_OFDM_M10000,1);
RkBar2 = abs(RkBar).^2;

Sim_M10000_avg = squeeze(mean(RkBar2,2)).';
Sim_M10000_avg = Sim_M10000_avg/max(Sim_M10000_avg)+1e-14;
Sim_M10000_avg = fftshift(Sim_M10000_avg);

%% Plot Fig. 6(a)
x = -N*L/2:N*L/2-1;

figure('Color','w','Position',[100,100,760,620]);
hold on;

plot(x,10*log10(Sim_M1_avg),'-', 'Color',[0,0.4470,0.7410],'LineWidth',1.2, 'DisplayName','No Integration, Numerical');
plot(x,10*log10(TheoAveACF_OFDM_M1),':', 'Color',[0,0.4470,0.7410],'LineWidth',1.8, 'DisplayName','No Integration, Theoretical');

plot(x,10*log10(Sim_M10000_avg),'-', 'Color',[0.8500,0.3250,0.0980],'LineWidth',1.2, 'DisplayName','10k Coh Integration, Numerical');
plot(x,10*log10(TheoAveACF_OFDM_M10000),':', 'Color',[0.8500,0.3250,0.0980],'LineWidth',1.8, 'DisplayName','10k Coh Integration, Theoretical');

plot(x,10*log10(TheoAveACF_Iceberg),'k--','LineWidth',1.2, 'DisplayName','"Iceberg" of the Designed Pulse');

xlabel('Delay Index');
ylabel('Ambiguity Level (dB)');
xlim([-300,300]);
ylim([-120,0]);
xticks(-300:50:300);
yticks(-120:20:0);
grid on;
box on;
legend('Location','south','EdgeColor','black');
set(gca,'FontName','Times New Roman','FontSize',14);

exportgraphics(gcf,'Fig6_a.png','Resolution',300);
exportgraphics(gcf,'Fig6_a.pdf','ContentType','vector');

%% ===========================================
width = 8;%设置图宽，这个不用改
height = 4;%设置图高，这个不用改
fontsize = 14;%设置图中字体大小
linewidth = 2;%设置线宽，一般大小为2，好看些。1是默认大小
markersize = 10;%标记的大小，按照个人喜好设置。
set(groot, 'defaultAxesFontName', 'Times New Roman');
set(groot, 'defaultTextFontName', 'Times New Roman');
set(groot, 'defaultLegendFontName', 'Times New Roman');
% ===========================================
figure(1);
% gca表示对axes的设置；  gcf表示对figure的设置
set(gcf, 'Units', 'inches');
% set(gcf, 'Position', [0, 0, width, height]);
set(gcf, 'Color', 'white'); % 设置背景是白色的 原先是灰色的 论文里面不好看
set(gcf, 'Renderer', 'painters');
set(gcf, 'PaperUnits', 'inches');
set(gcf, 'PaperPosition', [0, 0, width, height]);
set(gcf, 'PaperSize', [width, height]);

plot(x,10*log10(Sim_M1_avg),'-', 'Color','#F65314', 'LineWidth', 2); hold on;
plot(x,10*log10(TheoAveACF_OFDM_M1),':', 'Color', '#F65314', 'LineWidth', 2);
plot(x,10*log10(Sim_M10000_avg),'-', 'Color','#00A1F1', 'LineWidth', 2);
plot(x,10*log10(TheoAveACF_OFDM_M10000),':', 'Color', '#00A1F1', 'LineWidth', 1.5);
plot(x,10*log10(TheoAveACF_Iceberg),'k--', 'Color', '#B0B0B0', 'LineWidth', 1.5);


% 设置坐标轴的数字大小，包括xlabel/ylabel文字(坐标轴标注)大小.同时影响图例、标题等,除非它们被单独设置。
% 所以一开始就使用这行先设置刻度字体字号，然后在后面在单独设置坐标轴标注、图例、标题等的 字体字号。
set(gca, 'FontSize',16,'FontName','Times New Roman');
h_legend =  legend('No Integration, Numerical', 'No Integration, Theoretical', '10k Coh Integration, Numerical', '10k Coh Integration, Theoretical', '"Iceberg" of the Designed Pulse','Interpreter', 'latex');
legendsize = 13;
set(h_legend,'FontName','Times New Roman','FontSize',legendsize,'FontWeight','normal','LineWidth',1,'Location','northwest');
% set(h_legend,'Interpreter','latex') %  'box','off');
% h_legend.Interpreter = 'latex';
labelsize = 16;
xlabel('Delay Index', 'FontSize', labelsize, 'FontName', 'Times New Roman', 'Interpreter', 'latex');
ylabel('Ambiguity Level (dB)', 'FontSize', labelsize, 'FontName', 'Times New Roman', 'Interpreter', 'latex');

xlim([-300,300]);
ylim([-120,0]);
xticks(-300:50:300);
yticks(-120:20:0);
%----- Grid 设置----------------
grid on;
set(gca,'GridLineStyle', '--', 'Gridalpha',0.2, 'LineWidth', 1, 'GridLineWidth', 0.5, 'Layer','bottom');

%--------- savefig-------------
set(gca, 'Units', 'normalized');
set(gca, 'Position', [0.11, 0.12, 0.87, 0.86]);

% print(gcf, 'Fig6_a.pdf', '-dpdf', '-vector');






%%  
function g_opt = solve_iceberg_shaping_psl(N,L,alpha,K_s1)
N_alpha = fix(alpha*N);
N_non_rolloff = N-N_alpha;
N_zeros = floor(N_non_rolloff/2);
N_ones = floor(N_non_rolloff/2);

cvx_begin quiet
    variable g(N) nonnegative
    expressions psl_terms(length(K_s1))

    for indexK = 1:length(K_s1)
        k = K_s1(indexK);
        f_k = exp(-1j*2*pi*k*(0:N-1).'/(L*N));
        gk = g+(1-g)*exp(-1j*2*pi*k/L);
        psl_terms(indexK) = square_abs(f_k'*gk);
    end

    minimize(max(psl_terms))

    subject to
        % Reversed FFT-bin ordering used by the numerical implementation:
        % 1 -> monotonically decreasing roll-off -> 0.
        g(1:N_zeros) == 1;
        g(N-N_ones+1:N) == 0;

        for n = 1:N-1
            g(n+1)-g(n) <= 0;
        end

        sum(g) == N/2;
cvx_end

if strcmp(cvx_status,'Solved') || strcmp(cvx_status,'Inaccurate/Solved')
    fprintf('Optimization successful!\n');
    fprintf('Optimal PSL value: %.12e\n',cvx_optval);
    g_opt = g;
else
    error('CVX failed to solve the PSL problem. Status: %s',cvx_status);
end
end


function mat = FFTmatrix(L)
mat = complex(zeros(L,L));
for i = 0:L-1
    for j = 0:L-1
        mat(i+1,j+1) = exp(-1j*2*pi*i*j/L)/sqrt(L);
    end
end
end